SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialDebugger.cpp

```diff
@@ -27,10 +27,174 @@
 #include "Modules/ModuleManager.h"
 #include "Net/UnrealNetwork.h"
 
+#if UE_EDITOR
+#include "Editor.h"
+#endif
+
 using namespace SpatialGDK;
 
 DEFINE_LOG_CATEGORY(LogSpatialDebugger);
 
+FSpatialDebuggerSystem::FSpatialDebuggerSystem(USpatialNetDriver* InNetDriver, const FSubView& InSubView)
+	: NetDriver(InNetDriver)
+	, SubView(&InSubView)
+{
+	check(IsValid(InNetDriver));
+
+	EntityActorMapping.Reserve(ENTITY_ACTOR_MAP_RESERVATION_COUNT);
+}
+
+FSpatialDebuggerSystem::~FSpatialDebuggerSystem() = default;
+
+void FSpatialDebuggerSystem::AdvanceView()
+{
+	for (TMap<Worker_EntityId_Key, TWeakObjectPtr<AActor>>::TIterator It = EntityActorMapping.CreateIterator(); It; ++It)
+	{
+		if (!It->Value.IsValid())
+		{
+			It.RemoveCurrent();
+		}
+	}
+
+	for (const EntityDelta& EntityDelta : SubView->GetViewDelta().EntityDeltas)
+	{
+		switch (EntityDelta.Type)
+		{
+		case EntityDelta::ADD:
+			OnEntityAdded(EntityDelta.EntityId);
+			break;
+		case EntityDelta::REMOVE:
+			OnEntityRemoved(EntityDelta.EntityId);
+			break;
+		case EntityDelta::TEMPORARILY_REMOVED:
+			OnEntityRemoved(EntityDelta.EntityId);
+			OnEntityAdded(EntityDelta.EntityId);
+			break;
+		default:
+			break;
+		}
+
+		for (const AuthorityChange& AuthorityChange : EntityDelta.AuthorityGained)
+		{
+			if (AuthorityChange.Type == AuthorityChange::AUTHORITY_GAINED
+				&& AuthorityChange.ComponentSetId == SpatialConstants::SERVER_AUTH_COMPONENT_SET_ID)
+			{
+				ActorAuthorityGained(EntityDelta.EntityId);
+			}
+		}
+	}
+}
+
+void FSpatialDebuggerSystem::OnEntityAdded(const Worker_EntityId EntityId)
+{
+	check(NetDriver != nullptr);
+	if (NetDriver->IsServer())
+	{
+		if (SubView->HasAuthority(EntityId, SpatialConstants::SERVER_AUTH_COMPONENT_SET_ID))
+		{
+			ActorAuthorityGained(EntityId);
+		}
+
+		return;
+	}
+
+	TWeakObjectPtr<AActor>* ExistingActor = EntityActorMapping.Find(EntityId);
```

Comment:
why do we need this check? Is there a reason the subview wouldn't only give you adds for entities that don't currently exist in view?